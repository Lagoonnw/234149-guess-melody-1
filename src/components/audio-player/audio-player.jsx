import React, {PureComponent, createRef} from 'react';
import PropTypes from 'prop-types';
import {PlayButtonText} from '../../constants/constants';

export class AudioPlayer extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      progress: 0,
      isLoading: true,
      isPlaying: props.isPlaying,
    };
    this._audioRef = createRef();
    this._handleButtonClick = this._handleButtonClick.bind(this);
  }

  componentDidMount() {
    const {src} = this.props;
    const audio = this._audioRef.current;
    this._initAudioPlayer({audio, src});
  }

  componentDidUpdate() {
    const audio = this._audioRef.current;

    if (this.props.isPlaying) {
      audio.play();
    } else {
      audio.pause();
    }
  }

  componentWillUnmount() {
    const audio = this._audioRef.current;
    this._resetAudioPlayer(audio);
  }

  render() {
    const {src, controls} = this.props;
    const {isLoading, isPlaying} = this.state;

    return (
      <React.Fragment>
        <button
          className={`track__button track__button--${isPlaying ? PlayButtonText.PAUSE : PlayButtonText.PLAY}`}
          type="button"
          disabled={isLoading}
          onClick={this._handleButtonClick}
        />
        <div className="track__status">
          <audio
            src={src}
            controls={controls}
            ref={this._audioRef}
          />
        </div>
      </React.Fragment>
    );
  }

  _initAudioPlayer({audio, src}) {
    audio.src = src;

    audio.oncanplaythrough = () => this.setState({
      isLoading: false,
    });

    audio.onplay = () => {
      this.setState({
        isPlaying: true,
      });
    };

    audio.onpause = () => this.setState({
      isPlaying: false,
    });

    audio.ontimeupdate = () => this.setState({
      progress: audio.currentTime
    });
  }

  _resetAudioPlayer(audio) {
    audio.oncanplaythrough = null;
    audio.onplay = null;
    audio.onpause = null;
    audio.ontimeupdate = null;
    audio.src = ``;
  }

  _handleButtonClick() {
    this.props.onPlayButtonClick();
    this.setState({isPlaying: !this.state.isPlaying});
  }
}

AudioPlayer.propTypes = {
  isPlaying: PropTypes.bool,
  src: PropTypes.string.isRequired,
  onPlayButtonClick: PropTypes.func.isRequired,
  controls: PropTypes.bool
};

AudioPlayer.defaultProps = {
  isPlaying: false,
  controls: false
};


