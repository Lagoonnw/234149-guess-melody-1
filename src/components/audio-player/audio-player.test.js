import React from 'react';
import renderer from 'react-test-renderer';
import {AudioPlayer} from "./audio-player.jsx";

function createNodeMock(element) {
  if (element.type === `audio`) {
    return {};
  }
  return null;
}

const mock = {
  src: `bla-bla.mp3`,
  onPlayButtonClick: jest.fn()
};

describe(`Audio-player snapshot test`, () => {
  test(`Should render player correctly`, () => {
    const options = {createNodeMock};
    const tree = renderer.create(<AudioPlayer {...mock}/>, options).toJSON();
    expect(tree).toMatchSnapshot();
  });
});


