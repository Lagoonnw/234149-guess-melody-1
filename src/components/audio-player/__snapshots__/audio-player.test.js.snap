// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audio-player snapshot test Should render player correctly 1`] = `
Array [
  <button
    className="track__button track__button--play"
    disabled={true}
    onClick={[Function]}
    type="button"
  />,
  <div
    className="track__status"
  >
    <audio
      controls={false}
      src="bla-bla.mp3"
    />
  </div>,
]
`;
